File: src/main/java/hackerrank/graphs/FindTheNearestClone.java
package hackerrank.graphs;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public class FindTheNearestClone {

    static int findShortest(
            int graphNodes,
            int[] graphFrom,
            int[] graphTo,
            long[] ids,
            int val) {

        final CNode[] nodes = generateNodes(ids, graphFrom, graphTo);
        return findShortestColorMatchingDistance(nodes[val - 1]);
    }

    private static CNode[] generateNodes(
            long[] ids,
            int[] graphFrom,
            int[] graphTo) {

        final CNode[] nodes = new CNode[ids.length];
        for (int i = 0; i < ids.length; i++) {
            nodes[i] = new CNode(ids[i]);
        }

        for (int i = 0; i < graphFrom.length; i++) {
            final CNode from = nodes[graphFrom[i] - 1];
            final CNode to = nodes[graphTo[i] - 1];
            from.addNeighbors(to);
        }

        return nodes;
    }

    private static int findShortestColorMatchingDistance(CNode start) {
        final Queue<CNode> queue = new ArrayDeque<>();
        queue.offer(start);
        start.distance = 0;

        while (!queue.isEmpty()) {
            final CNode polled = queue.poll();

            for (CNode n : polled.neighbors) {
                if (n.isNotVisited()) {
                    n.distance = polled.distance + 1;
                    if (start.color.equals(n.color)) return n.distance;
                    queue.offer(n);
                }
            }
        }

        return -1;
    }

    static class CNode {

        private final Long color;
        Integer distance = -1;
        private final List<CNode> neighbors = new ArrayList<>();

        CNode(Long color) {
            this.color = color;
        }

        void addNeighbors(CNode neighbor) {
            if (neighbors.contains(neighbor)) return;
            neighbors.add(neighbor);
            neighbor.neighbors.add(this);
        }

        boolean isNotVisited() {
            return distance.equals(-1);
        }

    }

}

File: src/main/kotlin/hackerrank/graphs/FindTheNearestClone.kt
package hackerrank.graphs

import java.util.*

fun findShortest(
    graphNodes: Int,     // an integer, the number of nodes
    graphFrom: IntArray, // an array of integers, the start nodes for each edge
    graphTo: IntArray,   // an array of integers, the end nodes for each edge
    ids: LongArray,      // an array of integers, the color id per node
    `val`: Int           // an integer, the id of the color to match
): Int {

    val nodes = generateNodes(ids, graphFrom, graphTo)
    return findShortestColorMatchingDistance(nodes[`val` - 1])
}

private fun generateNodes(
    ids: LongArray,
    graphFrom: IntArray,
    graphTo: IntArray
): Array<CNode> {
    val nodes: Array<CNode> = ids
        .mapIndexed { index, color -> CNode(index + 1, color) }
        .toTypedArray()

    graphFrom.indices.forEach {
        val from = nodes[graphFrom[it] - 1]
        val to = nodes[graphTo[it] - 1]
        from.addNeighbor(to)
    }

    return nodes
}

private fun findShortestColorMatchingDistance(start: CNode): Int {
    val queue: Queue<CNode> = ArrayDeque<CNode>()
    queue.offer(start)
    start.distance = 0

    while (queue.isNotEmpty()) {
        val polled = queue.poll()

        polled.neighbors.forEach { n ->
            if (n.isNotVisited()) {
                n.distance = polled.distance + 1
                if (start.matchColor(n)) return n.distance
                queue.offer(n)
            }
        }
    }

    return -1
}

internal data class CNode(
    val value: Int,
    val color: Long,
    var distance: Int = initialDistance,
    val neighbors: MutableSet<CNode> = mutableSetOf()
) {

    companion object {
        const val initialDistance = -1
    }

    fun addNeighbor(node: CNode) {
        this.neighbors.add(node)
        node.neighbors.add(this)
    }

    fun isNotVisited() = distance == initialDistance

    fun matchColor(compare: CNode) = color == compare.color

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as CNode
        return value == other.value
    }

    override fun hashCode(): Int {
        return value
    }

    override fun toString(): String {
        return "CNode(value=$value, color=$color, distance=$distance)"
    }

}